core/query/src/main/java/org/eclipse/rdf4j/query/Query.java

```diff
@@ -44,21 +44,44 @@
 	/**
 	 * Specifies the maximum time that a query is allowed to run. The query will be interrupted when it exceeds the time
 	 * limit. Any consecutive requests to fetch query results will result in {@link QueryInterruptedException}s.
-	 * 
+	 *
 	 * @param maxQueryTime The maximum query time, measured in seconds. A negative or zero value indicates an unlimited
 	 *                     query time (which is the default).
 	 * @deprecated since 2.0. Use {@link Operation#setMaxExecutionTime(int)} instead.
 	 */
 	@Deprecated
-	public void setMaxQueryTime(int maxQueryTime);
+	void setMaxQueryTime(int maxQueryTime);
 
 	/**
 	 * Returns the maximum query evaluation time.
-	 * 
+	 *
 	 * @return The maximum query evaluation time, measured in seconds.
 	 * @see #setMaxQueryTime(int)
 	 * @deprecated since 2.0. Use {@link Operation#getMaxExecutionTime()} instead.
 	 */
 	@Deprecated
-	public int getMaxQueryTime();
+	int getMaxQueryTime();
+
+	// TODO - make default with no-op before merging to develop for backwards compatibility
+	// Also wondering if it makes sense to move this up a level to the Operation interface? Since we will be using a
+	// default no-op implementation we could move it without affecting backwards compatibility.
+	QueryExplainWrapper explain(QueryExplainLevel queryExplainLevel);
+
+	/**
+	 * The different types of queries that RDF4J recognizes: boolean queries, graph queries, and tuple queries.
+	 *
+	 * @since 3.2.0
+	 */
+	enum QueryExplainLevel {
+		Unoptimized,
+		Optimized,
+		Executed
+	}
+
+	interface QueryExplainWrapper {
+
+		String asHumanReadbleString();
+		// TupleExpr asTupleExpr(); location in maven hierarchy prevents us from using TupleExpr here
```

Comment:
I’m trying to consider what we could do in the future to give advanced users more control of their queries. 

An example is a query I have which takes hours to complete, but it’ll easily run in milliseconds if certain query optimizers are turned off. 

Other databases have query hints that users can add to their sparql queries for this purpose. 

I was playing with the idea of exposing the tupleexpr to facilitate some similar, but on a lower level. 

Also been thinking about using the resultSizeActual for optimizing a query that gets run frequently.

Reply:
Sounds interesting, but I think we're getting a bit far afield here from the scope. However, I can see that potentially tying it in more closely with TupleExpr is something we might want to keep our options open about. A simple solution would be to move this QueryExplanation interface into the query.algebra package (as part of the rdf4j-queryalgebra-model module). It's defensible to put it there anyway I think, and that way we will have the option in future to expose relevant bits of the query algebra model via this interface.

Reply:
<img width="425" alt="Screenshot 2020-04-25 at 13 36 29" src="https://user-images.githubusercontent.com/797185/80278830-c77c1d00-86f9-11ea-83a7-f730314947bd.png">

Reply:
I cant move the interface from the `query` package to the `queryalgebra` package because that would cause this circular dependency. I haven't looked into it yet though.